### PR TITLE
Added mean algorithm and test code

### PR DESCRIPTION
**Pull request by Jason07311**
Hello I would like to contribute to an algorithm that computes mean related in Math (Statistic). Thank you very much!
feat: added mean.go
test: added mean_test.go with 4 test cases
docs: added comment and explanation to mean.go & mean_test.go

### CODE REVIEW

File: math/mean/mean.go [hunk 1]
```diff
@@ -0,0 +1,27 @@
+package mean
```
Comment (tjgurwara99) [inline math/mean/mean.go:1]: You do not need to create a separate package just for one function. This clearly belongs in the `math` package.

File: math/mean/mean.go [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+
+import (
+	"fmt"
+)
+
+func mean(values []float64) float64{
+
+	if len(values) <= 0 {
+		fmt.Println("No values detected")
```
Comment (tjgurwara99) [inline math/mean/mean.go:10]: No need for this.

File: math/mean/mean.go [hunk 3]
```diff
@@ -0,0 +1,27 @@
+package mean
+
+import (
+	"fmt"
+)
+
+func mean(values []float64) float64{
```
Comment (tjgurwara99) [inline math/mean/mean.go:7]: Please capitalise this function, otherwise how would you expect it to work outside this package.

File: math/mean/mean_test.go [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+		{
+			name: "With negative values",
+			testValues: []float64{-1,2,-3,4,5},
+			average: 1.4,
+		},
+		{
+			name: "With floating values",
+			testValues: []float64{1.1,2.2,3.3,4.4,5.5},
+			average: 3.3,
+		},
```
Comment (tjgurwara99) [inline math/mean/mean_test.go:32]: A another test case with no values inside.

```
testValues: []float64{},
```

to make sure you cover this edge case.

Comment (tjgurwara99) [inline math/mean/mean_test.go:32]:
```suggestion
		},
		{
 			name: "With no values",
 			testValues: []float64{},
 			average: 0,
 		},
```

File: math/mean/mean.go (hunk 3 above)
Comment (tjgurwara99) [inline math/mean/mean.go:7]: Please also consider using generics here, this is clearly something that would benefit from generics.

**Review by tjgurwara99 — CHANGES_REQUESTED** [review]
Please consider the suggestions provided.

File: math/mean/mean.go (hunk 2 above)
Comment (Jason07311) [reply math/mean/mean.go:10]: Sorry to ask, is the meaning of "No need" referring to removing the whole validating condition of empty array ? Or just the line of code printing the message ?

One reason of not suggesting to remove the validating condition above is to consider the case where  empty array is passed into the function.

File: math/mean/mean.go (hunk 3 above)
Comment (Jason07311) [reply math/mean/mean.go:7]: Just added the generics. Please review.

File: math/mean/mean_test.go (hunk 4 above)
Comment (Jason07311) [reply math/mean/mean_test.go:32]: Just added the new test case. Please review.

File: math/mean/mean.go (hunk 3 above)
Comment (Jason07311) [reply math/mean/mean.go:7]: Just edited the function name. Please review.

File: math/mean/mean.go (hunk 1 above)
Comment (Jason07311) [reply math/mean/mean.go:1]: Just moved the mean.go and the mean_test.go under the directory of maths

File: math/mean/mean.go (hunk 2 above)
Comment (tjgurwara99) [reply math/mean/mean.go:10]: Just the line printing the message.

File: math/mean/mean.go (hunk 1 above)
Comment (tjgurwara99) [reply math/mean/mean.go:1]: I can still see the file present. You need to remove this file and commit the removal to git and push your changes. Otherwise it will remain in the source control.

File: math/mean.go [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+	"golang.org/x/exp/constraints"
+)
+
+type Digit interface {
+	constraints.Float | constraints.Integer
+}
+
+func Mean[T Digit](values []T) float64 {
+
+	// Validating the empty array
```
Comment (tjgurwara99) [inline math/mean.go:13]: No need for these kind of comments. They usually don't provide value.

When adding comments to code, there are usually two things that you should consider:

1. are you doing something that is very complex that will not be understood by someone reading your code.
2. are you explain why you're doing something rather than what you're doing.

If none of the above cases apply then it is (usually) just considered noise.

File: math/mean/mean.go (hunk 2 above)
Comment (Jason07311) [reply math/mean/mean.go:10]: > Just the line printing the message.

Got it, just removed it in my latest committed work

File: math/mean.go [hunk 6]
```diff
@@ -0,0 +1,31 @@
...
+	if len(values) == 0 {
+		return 0
+	}
+
+	var summation float64 = 0
+
+	// Summation of all values in the slice
```
Comment (tjgurwara99) [inline math/mean.go:20]: Same here.

File: math/mean.go [hunk 7]
```diff
@@ -0,0 +1,31 @@
...
+	for _, singleValue := range values {
+		summation += float64(singleValue)
+	}
+
+	// Casting to float64 done for the len()
```
Comment (tjgurwara99) [inline math/mean.go:25]: same here.

File: math/mean.go [hunk 8]
```diff
@@ -0,0 +1,31 @@
...
+	var average = summation / float64(len(values))
+
+	// Return of average value
```
Comment (tjgurwara99) [inline math/mean.go:28]: Same here.

File: math/mean/mean_test.go [hunk 9]
```diff
@@ -0,0 +1,47 @@
+package mean
```
Comment (tjgurwara99) [inline math/mean/mean_test.go:1]: You need to remove this file too.

**Review by tjgurwara99 — CHANGES_REQUESTED** [review]
Please consider the suggestion below/

**Comment by tjgurwara99** [comment]
Also, you still haven't run `gofmt` on your code. It will keep failing our automated tests until you do.

File: math/mean.go [hunk 10]
```diff
@@ -0,0 +1,31 @@
+package maths
```
Comment (tjgurwara99) [inline math/mean.go:1]:
```suggestion
package math
```

File: math/mean.go [hunk 11]
```diff
@@ -0,0 +1,31 @@
...
+
+import (
+	"golang.org/x/exp/constraints"
+)
+
+type Digit interface {
```
Comment (tjgurwara99) [inline math/mean.go:7]: We already have those [constraints](https://github.com/TheAlgorithms/Go/blob/master/constraints/contraints.go#L28-L30)

File: math/mean.go [hunk 12]
```diff
@@ -0,0 +1,31 @@
...
+
+	var summation float64 = 0
+
+	// Summation of all values in the slice
+	for _, singleValue := range values {
+		summation += float64(singleValue)
+	}
+
+	// Casting to float64 done for the len()
+	var average = summation / float64(len(values))
```
Comment (tjgurwara99) [inline math/mean.go:26]:
```suggestion
	return summation / float64(len(values))
```

Comment (tjgurwara99) [reply math/mean.go:26]: With this, everything after this is redundant.

**Review by tjgurwara99 — CHANGES_REQUESTED** [review]
Please fix these issues.

File: math/mean.go [hunk 13]
```diff
@@ -0,0 +1,25 @@
+package math
+
+import (
+	"github.com/TheAlgorithms/Go/constraints"
+)
+
+type Digit interface {
```
Comment (tjgurwara99) [inline math/mean.go:7]: This is useless if you're using the `constraints` package.

We already have `constraints.Number` which is exactly the same as this. Please use that instead and remote this.

File: math/mean_test.go [hunk 14]
```diff
@@ -0,0 +1,51 @@
+package math
```
Comment (tjgurwara99) [inline math/mean_test.go:1]:
```suggestion
package math_test
```

**Review by tjgurwara99 — CHANGES_REQUESTED** [review]

File: math/mean.go (hunk 13 above)
Comment (Jason07311) [reply math/mean.go:7]: Ya, I forgot to remove them in my last push. Sorry for that.

File: math/mean_test.go (hunk 14 above)
Comment (Jason07311) [reply math/mean_test.go:1]: Does this indicate of creating another directory for math_test ?

Comment (Jason07311) [reply math/mean_test.go:1]: > Does this indicate of creating another directory for math_test ?

Or just moving the whole test script to the math_test.go under math package?

Comment (Jason07311) [reply math/mean_test.go]: Done updating the package.

**Review by tjgurwara99 — APPROVED** [review]

**Review by raklaptudirm — APPROVED** [review]

**Comment by Jason07311** [comment]
Thanks for the effort upon reviewing my contribution and providing the suggestions, really appreciate that.
